core/app/app.go

```diff
@@ -0,0 +1,213 @@
+package app
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	format "github.com/ipfs/go-ipld-format"
+	logging "github.com/ipfs/go-log"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	"github.com/textileio/go-threads/broadcast"
+	"github.com/textileio/go-threads/core/net"
+	"github.com/textileio/go-threads/core/thread"
+)
+
+var log = logging.Logger("app")
+
+const (
+	busTimeout        = time.Second * 10
+	addRecordTimeout  = time.Second * 10
+	fetchEventTimeout = time.Second * 15
+)
+
+// App provides a bidirectional hook for thread-based apps.
+type App interface {
```

Comment:
An app implements these two methods. Basically, it must provide a hook for threads to handle new events, and it must itself handle new records.